File: backend/src/main/java/com/roccia/backend/dto/ErrorResponse.java
package com.roccia.backend.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Builder;
import lombok.Getter;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@Builder
public class ErrorResponse {
    @Builder.Default // builder에 명시되어있지 않으면 null로 들어가는거 방어
    private final LocalDateTime timestamp = LocalDateTime.now();
    private final int status;
    private final String error;
    private final String message;

    @JsonInclude(JsonInclude.Include.NON_NULL) // null인 필드는 JSON 응답에서 아예 제외 (의미가 헷갈리지 않게)
    private final List<FieldError> errors;

    @Getter
    @Builder
    public static class FieldError {
        private final String field;
        private final String value;
        private final String reason;
    }
}

File: frontend/lib/models/error_response.dart
class ErrorResponse {
  final String? timestamp;
  final int status;
  final String error;
  final String message;
  final List<FieldError>? errors;

  ErrorResponse({
    this.timestamp,
    required this.status,
    required this.error,
    required this.message,
    this.errors,
  });

  factory ErrorResponse.fromJson(Map<String, dynamic> json) {
    return ErrorResponse(
      timestamp: json['timestamp'] as String?,
      status: json['status'] as int,
      error: json['error'] as String,
      message: json['message'] as String,
      errors: json['errors'] != null
          ? (json['errors'] as List).map((i) => FieldError.fromJson(i)).toList()
          : null,
    );
  }
}

class FieldError {
  final String field;
  final String value;
  final String reason;

  FieldError({
    required this.field,
    required this.value,
    required this.reason,
  });

  factory FieldError.fromJson(Map<String, dynamic> json) {
    return FieldError(
      field: json['field'] as String,
      value: json['value']?.toString() ?? '',
      reason: json['reason'] as String,
    );
  }
}
